### ISSUE TITLE
Form Submission e2e Test Improvement

### ISSUE THREAD

**Issue by braughtg** [FarmData2/FarmData2#170]
Currently the unit tests for the `submit` function in `lib.js` for each entry point (Tray Seeding / Direct Seeding) thoroughly test that the correct assets, quantities and logs get created in the farmOS database based on the `form` that they are passed.

In normal use, the `form` is created by the UI and passed to the `submit` function.  Currently, the e2e tests for each entry point test that the `submit` function is called by intercepting the final API request that it makes.  They do not however test that all of the assets, quantities and logs are created.  They should not have to, because `submit` is fully tested by the unit tests.  

However, these e2e tests are currently unable to efficiently ensure that all of the correct information is in the form that is passed to `submit`.  The best way to do this would be to _spy_ on the `submit` function and simply check the contents of the argument that is passed. However, I have not been able to get a spy to work on this function due to import issues.  More specifically, the test would need to import `lib.js` which it can do.  However, `lib.js` imports `@libs/farmosUtil/farmosUtil` where `@libs` is defined in the `vite.config.js` file, and for some reason that import is failing to resolve.

It would be good to have some efficient way to test that `submit` is getting all of the information that is is supposed to get.  Take _efficient_ here to mean not asking the database for all of the assets, quantities and logs.

It seems like there should be a good way to do this with `cy.spy` but the build issues are currently a show stopper.

When a solution has been found it should be back ported to:
- [x] - Tests for tray Seeding `lib.js`
- [x] - Tests for Direct Seeding `lib.js`
- [x] - Tests for Transplanting `lib.js` - if it has been written.

**Comment by braughtg** [comment]
Should be able to use the `intercept`. Set an intercept for each of the API requests that are used to make a farmOS asset, log or quantity.  Then use the intercept to grab the payload of the `POST` request.  Then the test can check the contents of that payload.

Something like this: https://docs.cypress.io/api/commands/intercept#Asserting-on-a-request

Some general resources that may be of help:
- https://docs.cypress.io/guides/guides/network-requests
- https://docs.cypress.io/api/commands/intercept

**Comment by braughtg** [comment]
It might be preferable to actually `stub` the `submitForm` method.  Then there could be a separate test to check the content of the `form` that does not do a submit.  Then there could be separate tests that check for the `toast` messages.

Something similar to https://github.com/cypress-io/cypress/issues/22355#issuecomment-1485716724 might be a path. They say:

This, which is what we are doing now, will not work:
```
/// module.js
export function myFunc() {}

// module.spec.js
import * as MyModule from './module.js'
cy.stub(MyModule, 'myFunc')
```

But they say that this will work:
```Javascript
/// module.js
function myFunc() {}
export const MyModule = {
    myFunc
}

// module.spec.js
import { MyModule } from './module.js'
cy.stub(MyModule, 'myFunc')
```

**Comment by braughtg** [comment]
It is seeming that this is not going to work for e2e testing as the `lib` that is imported in the test is separate from the `lib` that is imported in the `App.vue`.  The following is the confirmation:

A test with the following structure fails:
```JavaScript
cy.spy(lib, 'submitForm').as('submitFuncSpy');
...
// Clicking the button calls lib.submitForm() in the page
cy.get('[data-cy="submit-button"]').click(); 
...
cy.get('@submitFuncSpy').should('be.called');
...
```

While one with this structure passes:
```JavaScript
cy.spy(lib, 'submitForm').as('submitFuncSpy');
...
// Clicking the button calls lib.submitForm() in the page
cy.get('[data-cy="submit-button"]').click(); 

// Call lib.submitForm() explicitly in the test
lib.submitForm(null).catch(() => {});
...
cy.get('@submitFuncSpy').should('be.called');
...
```

**Comment by braughtg** [comment]
> It is seeming that this is not going to work for e2e testing as the `lib` that is imported in the test is separate from the `lib` that is imported in the `App.vue`.

One additional path to pursue is to see if a global (`window`) variable can be created in the `App.vue` that is then accessed in the `*.e2e.cy.js` file.  Then the `lib.submitForm` function can be assigned to that variable and could be spied from the test.

**Comment by braughtg** [comment]
> One additional path to pursue is to see if a global (`window`) variable can be created in the `App.vue` that is then accessed in the `*.e2e.cy.js` file.

The following is working:

1. Change `lib.js` to not export the `submitForm` function directly but instead to use the following at the end of the file: 
   ```JavaScript
   export const lib = {
     submitForm,
   };
   ```

2. Change the `lib` import in `App.vue`, `lib.submit.unit.cy.js` and `lib.submitError.unit.cy.js` from:
   ```JavaScript
     import * as lib from `./lib.js`
   ```
   to:
   ```JavaScript
   import { lib } from './lib.js';
   ```
   This brings `lib` in as a mutable object that will allow it to be assigned to a `window` variable and they spied by the test.

3. Change `App.vue` to:
   - Change the call to `lib.submitForm` to pass a copy of the `form` so that when the form is reset the passed values area still available in the test.
     ```JavaScript
      lib
        .submitForm({ ...this.form })
     ```
   - Add the following to the `created` lifecycle hook:
     ```JavaScript
     if (window.Cypress) {
       /* 
        * Make the lib containing the submitForm function accessible to the
        * e2e tests so that the submission test can spy on the submitForm 
        * function to verify that it is receiving the correct information.
        */ 
       document.defaultView.lib = lib;
     }
     ```

4. Then in the `*.submisson.e2e` test file:
   - setup a spy:
     ```JavaScript
     cy.window().then((win) => {
       cy.spy(win.lib, 'submitForm').as('submitFormSpy');
     });
     ```
   - Get the spy and make assertions about the parameter it was called with:
     ```JavaScript
     cy.get('@submitFormSpy').then((spy) => {
       expect(spy).to.be.calledOnce

       let formData = spy.getCall(0).args[0];
       expect(formData.seedingDate).to.equal('1950-01-02');
       expect(formData.cropName).to.equal('ARUGULA');
       expect(formData.locationName).to.equal('CHUAU');
       expect(formData.trays).to.equal(10);
       expect(formData.traySize).to.equal('50');
       expect(formData.seedsPerCell).to.equal(3);
       expect(formData.comment).to.equal('test comment');
     }
   ```

**Comment by braughtg** [comment]
Closed by #271
